src/programs/can_encoder_index_test.cpp

```diff
@@ -0,0 +1,130 @@
+/**
+ * \file
+ * \brief Simple test tool to check if the encoder index is detected.
+ *
+ * Runs a motor at a constant torque and prints a line whenever the encoder
+ * index is detected.
+ *
+ * \copyright Copyright (c) 2020 Max Planck Gesellschaft.
+ */
+#include <array>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <cmath>
+
+#include <real_time_tools/thread.hpp>
+#include <real_time_tools/spinner.hpp>
+
+#include <blmc_robots/blmc_joint_module.hpp>
+
+
+struct Params
+{
+    std::string can_port;
+    int motor_idx;
+    double torque;
+};
+
+
+void* foo(void *params)
+{
+    const std::string can_port = ((Params*)params)->can_port;
+    const int motor_idx = ((Params*)params)->motor_idx;
+    const double torque = ((Params*)params)->torque;
+
+    std::cout << "CAN port: " << can_port << std::endl;
+    std::cout << "Motor Index: " << motor_idx << std::endl;
+    std::cout << "Torque: " << torque << std::endl;
+
+    std::cout << "==============================" << std::endl;
+    std::cout << "Initialising..." << std::endl;
+
+    // setup can bus
+    auto can_bus = std::make_shared<blmc_drivers::CanBus>(can_port);
+
+    // set up motor board
+    auto motor_board =
+        std::make_shared<blmc_drivers::CanBusMotorBoard>(can_bus, 1000, 10);
+    motor_board->wait_until_ready();
+
+    std::shared_ptr<blmc_drivers::MotorInterface> motor =
+        std::make_shared<blmc_drivers::Motor>(motor_board, motor_idx);
+
+    constexpr double torque_constant_NmpA = 0.02;
+    constexpr double gear_ratio = 9.0;
+    constexpr double max_current_A = 2.0;
+    constexpr double one_motor_rotation_distance = 2.0 * M_PI / gear_ratio;
+
+    blmc_robots::BlmcJointModule joint_module(
+        motor, torque_constant_NmpA, gear_ratio, 0, false, max_current_A);
+
+
+    std::cout << "Start moving with constant torque" << std::endl;
+    real_time_tools::Spinner spinner;
+    spinner.set_period(0.001);
+    double last_index_position = joint_module.get_measured_index_angle();
+    unsigned int counter = 0;
```

Comment:
In theory you shouldn't allocate memory while in the real time thread.
So in your struct you can also add the shared pointer to all the classes you use and instanciate them in the main.
I leave you the choice to do it or not.

Reply:
Is it actually a problem when I allocate memory before entering the actual control loop?  Before sending the first command to the board, it should not matter if real-time constraints are met or not.

Reply:
When you start the thread you actually do a memlock over all the process.
The RT_PREEMPT page strongly suggest to avoid allocating memory after you spawn the thread.
Of course in this demo you are not allocating memory during the control loop which is the most crucial part.
But I would still show the "best" behavior in a demo.

Reply:
Oh, good to know.  Alright, I'll change it then.

Reply:
I changed it to avoid any dynamic memory allocation in the thread.